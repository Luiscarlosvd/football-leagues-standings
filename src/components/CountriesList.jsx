import { useDispatch, useSelector } from 'react-redux';
import { useEffect, useState } from 'react';
import { changeRegion } from '../redux/countries/countriesSlice';
import worldImg from '../assets/pngegg (4).png';
import { getCountriesInfo } from '../redux/countries/countriesSlice';
import CountryCard from './CountryCard';

const CountriesList = () => {
  const country = useSelector(state => state.countries);
  const [ valueSearch, setValueSearch ] = useState('');
  const dispatch = useDispatch();

  useEffect(() => {
    if(country.countries.length === 0){
      dispatch(getCountriesInfo());
    }
  }, [dispatch, country.countries.length]);

  const displayCountriesByRegion = country.countries.filter((countries) => {
    return countries.region === country.region}
  ).filter((country) => {
    if(valueSearch === '') {
      return country
    } else {
      return country.name.includes(valueSearch);
    }
  });

  const handleChange = (e) => {
    dispatch(changeRegion(e.target.value));
  }

  const handleSearchValue = (e) => {
    const value = e.target.value.charAt(0).toUpperCase() + e.target.value.slice(1).toLowerCase();
    setValueSearch(value);
  }

  const totalPopulation = () => {
    let total = 0;
    displayCountriesByRegion.map((country) => {
      total += country.population;
    })
    return total;
  }

  return (
    <main className='flex flex-col justify-center items-center'>
      {country.status === "Loading" && <div className='text-white'>Loading...</div>}
      {country.status === "fulfilled" && 
        <>
          <div>
            <img src={worldImg} alt="Icon world map" />
          </div>
          <div className="w-2/3">
            <input className="text-lato-400 search-bar text-white bg-black bg-opacity-30 mt-4 rounded-md p-1 px-3 w-full" type="search" onChange={handleSearchValue} placeholder="Search country"/>
          </div>
          <div className="flex text-lato-400 items-center gap-2 mt-4">
            <p className="text-white" htmlFor="region">Region</p>
            <select name="region" className="text-white rounded-md bg-opacity-25 bg-black p-1 pr-7 text-lato-400" onChange={handleChange}>
              <option value="Europe">Europe</option>
              <option value="Americas">Americas</option>
              <option value="Asia">Asia</option>
              <option value="Oceania">Oceania</option>
              <option value="Africa">Africa</option>
              <option value="Antarctic">Antarctic</option>
            </select>
          </div>
          <div className="text-center mt-4 text-white">
            <p>Total Population of {country.region}:</p>
            <p>{totalPopulation()}</p>
          </div>
          <div className="mt-6 w-full bg-black bg-opacity-10">
            <p className="ml-4 py-1 text-sm text-lato-400 text-white">STATS BY COUNTRY</p>
          </div>
          <section className='grid grid-cols-2 w-full bg-blue-800'>
            {displayCountriesByRegion.map((country) => (
              <CountryCard
                key={country.name}
                name={country.name}
                flag={country.flag}
                population={country.population}
              />
            ))}
          </section>
        </>
      }
    </main>
  )
}

export default CountriesList;
