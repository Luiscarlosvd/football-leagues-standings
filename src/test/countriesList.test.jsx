import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { describe, test, expect } from 'vitest';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import CountriesList from '../components/CountriesList';

const customMiddleware = [thunk]
const mockStore = configureMockStore(customMiddleware);

describe('CountriesList', () => {
  test('renders loading state', () => {
    // Arrange
    const store = mockStore({
      countries: {
        status: 'Loading',
        countries: [],
        region: 'Europe',
      },
    });

    // Act
    render(
      <Provider store={store}>
        <CountriesList />
      </Provider>
    );

    // Assert
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });
  test('renders countries list correctly', () => {
    // Arrange
    const store = mockStore({
        countries: {
          status: 'fulfilled',
          countries: [
            {
              name: 'Argentina',
              flag: 'https://restcountries.com/data/arg.svg',
              population: 44938712,
              region: 'Americas',
            },
            {
              name: 'Brazil',
              flag: 'https://restcountries.com/data/bra.svg',
              population: 206135893,
              region: 'Americas',
            },
            ],
          region: 'Americas',
        },
    });

    // Act
    const {container} = render(
      <Provider store={store}>
        <BrowserRouter>
            <CountriesList />
        </BrowserRouter>
      </Provider>
    );

    // Assert
    expect(container).toMatchSnapshot();
  });
});