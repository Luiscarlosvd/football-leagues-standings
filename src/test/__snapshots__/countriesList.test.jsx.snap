// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountriesList > renders countries list correctly 1`] = `
<div>
  <main
    class="flex flex-col justify-center items-center"
  >
    <div>
      <img
        alt="Icon world map"
        src="/src/assets/pngegg (4).png"
      />
    </div>
    <div
      class="w-2/3"
    >
      <input
        class="text-lato-400 search-bar text-white bg-black bg-opacity-30 mt-4 rounded-md p-1 px-3 w-full"
        placeholder="Search country"
        type="search"
      />
    </div>
    <div
      class="flex text-lato-400 items-center gap-2 mt-4"
    >
      <p
        class="text-white"
        for="region"
      >
        Region
      </p>
      <select
        class="text-white rounded-md bg-opacity-25 bg-black p-1 pr-7 text-lato-400"
        name="region"
      >
        <option
          value="Europe"
        >
          Europe
        </option>
        <option
          value="Americas"
        >
          Americas
        </option>
        <option
          value="Asia"
        >
          Asia
        </option>
        <option
          value="Oceania"
        >
          Oceania
        </option>
        <option
          value="Africa"
        >
          Africa
        </option>
        <option
          value="Antarctic"
        >
          Antarctic
        </option>
      </select>
    </div>
    <div
      class="text-center mt-4 text-white"
    >
      <p>
        Total Population of 
        Americas
        :
      </p>
      <p>
        251074605
      </p>
    </div>
    <div
      class="mt-6 w-full bg-black bg-opacity-10"
    >
      <p
        class="ml-4 py-1 text-sm text-lato-400 text-white"
      >
        STATS BY COUNTRY
      </p>
    </div>
    <section
      class="grid grid-cols-2 w-full bg-blue-800"
    >
      <div
        class="flex text-right flex-col bg-blue-700 relative px-2 py-4 gap-8 items-center border border-blue-800 border-opacity-10y- justify-between"
      >
        <a
          class="absolute right-2 top-2"
          href="/details/Argentina"
        >
          <svg
            class="text-3xl text-gray-200"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128,26A102,102,0,1,0,230,128,102.12,102.12,0,0,0,128,26Zm0,192a90,90,0,1,1,90-90A90.1,90.1,0,0,1,128,218Zm44.24-94.24a6,6,0,0,1,0,8.48l-32,32a6,6,0,0,1-8.48-8.48L153.51,134H88a6,6,0,0,1,0-12h65.51l-21.75-21.76a6,6,0,0,1,8.48-8.48Z"
            />
          </svg>
        </a>
        <div
          class="self-start ml-3"
        >
          <img
            alt="Country image flag"
            class="w-28 h-4-5 rounded-sm"
            src="https://restcountries.com/data/arg.svg"
          />
        </div>
        <div
          class="self-end mr-3"
        >
          <h2
            class="text-white text-lg text-lato-700"
          >
            Argentina
          </h2>
          <p
            class="text-white text-sm text-lato-300"
          >
            Population: 
            44938712
          </p>
        </div>
      </div>
      <div
        class="flex text-right flex-col bg-blue-700 relative px-2 py-4 gap-8 items-center border border-blue-800 border-opacity-10y- justify-between"
      >
        <a
          class="absolute right-2 top-2"
          href="/details/Brazil"
        >
          <svg
            class="text-3xl text-gray-200"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128,26A102,102,0,1,0,230,128,102.12,102.12,0,0,0,128,26Zm0,192a90,90,0,1,1,90-90A90.1,90.1,0,0,1,128,218Zm44.24-94.24a6,6,0,0,1,0,8.48l-32,32a6,6,0,0,1-8.48-8.48L153.51,134H88a6,6,0,0,1,0-12h65.51l-21.75-21.76a6,6,0,0,1,8.48-8.48Z"
            />
          </svg>
        </a>
        <div
          class="self-start ml-3"
        >
          <img
            alt="Country image flag"
            class="w-28 h-4-5 rounded-sm"
            src="https://restcountries.com/data/bra.svg"
          />
        </div>
        <div
          class="self-end mr-3"
        >
          <h2
            class="text-white text-lg text-lato-700"
          >
            Brazil
          </h2>
          <p
            class="text-white text-sm text-lato-300"
          >
            Population: 
            206135893
          </p>
        </div>
      </div>
    </section>
  </main>
</div>
`;
